### PR TITLE
Add QAOA optimization experiment

### PR DESCRIPTION
**Pull request by kevinsung**
After #36

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -1,4 +1,5 @@
 cirq-unstable==0.8.0.dev20200318154950
+scikit-learn
```
Comment (mpharrigan) [inline requirements.txt:2]: can you delmit this requirement with spaces and a comment

```python

# optimize
scikit-learn
```

File: recirq/optimize/mgd_test.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    rate = 1e-1
+    known_xs = [np.ones(5)]
+    known_ys = [10.0]
+    _ = model_gradient_descent(sum_of_squares,
+                               x0,
+                               sample_radius=sample_radius,
+                               n_sample_points=10,
+                               rate=rate,
+                               tol=1e-8,
+                               known_values=(known_xs, known_ys),
```
Comment (mpharrigan) [inline recirq/optimize/mgd_test.py:54]: pycharm is mad that the known `y` is a `float` and not an `ndarray`. Fix annotation in `model_gradient descent`?

File: recirq/optimize/minimize.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
+from typing import Callable, Optional, Tuple
+
+import numpy as np
+import scipy
```
Comment (mpharrigan) [inline recirq/optimize/minimize.py:4]:
```suggestion
import scipy.optimize
```

File: requirements.txt (hunk 1 above)
Comment (kevinsung) [reply requirements.txt]: Done.

File: recirq/optimize/mgd.py [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+from scipy.optimize import OptimizeResult
+from sklearn.linear_model import LinearRegression
+from sklearn.preprocessing import PolynomialFeatures
+from sklearn.pipeline import Pipeline
+
+
+def _get_least_squares_model_gradient(xs: List[np.ndarray],
+                                      ys: List[np.ndarray], xopt: np.ndarray,
+                                      linear_model: LinearRegression
+                                     ) -> np.ndarray:
```
[the remark this answers is not in the document]
Comment (kevinsung) [reply recirq/optimize/mgd.py]: Done.

File: recirq/optimize/mgd.py [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+import numpy as np
+import scipy
+from scipy.optimize.optimize import wrap_function
+from scipy.optimize import OptimizeResult
+from sklearn.linear_model import LinearRegression
+from sklearn.preprocessing import PolynomialFeatures
+from sklearn.pipeline import Pipeline
+
+
+def _get_least_squares_model_gradient(xs: List[np.ndarray],
```
[the remark this answers is not in the document]
Comment (kevinsung) [reply recirq/optimize/mgd.py]: Done.

File: recirq/optimize/mgd.py [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+    ])
+    shifted_xs = [x - xopt for x in xs]
+    model = model.fit(shifted_xs, ys)
+    fitted_coeffs = model.named_steps['linear_model'].coef_
+    n = len(xs[0])
+    linear_coeffs = np.array(fitted_coeffs[1:n + 1])
+    return linear_coeffs, model
+
+
+def _random_point_in_ball(n: int, radius: float) -> np.ndarray:
```
[the remark this answers is not in the document]
Comment (kevinsung) [reply recirq/optimize/mgd.py:60]: Done.

File: recirq/optimize/mgd.py [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+    This algorithm works well when you have an initial guess which is in the
+    convex neighborhood of a local optimum and you want to converge to that
+    local optimum. It's meant to be used when the function is stochastic.
+
+    Args:
+        f: The function to minimize.
+        x0: An initial guess.
+        args: Additional arguments to pass to the function.
+        rate: The learning rate for the gradient descent.
+        sample_radius: The radius around the current iterate to sample
```
[the remark this answers is not in the document]
Comment (kevinsung) [reply recirq/optimize/mgd.py:108]: That's an interesting point. Yes, I think this optimizer should behave better when the landscape is not too "stretched out" in certain directions. I think the option you mention is reasonable to add, but perhaps in a separate PR.

File: recirq/optimize/mgd.py [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+            points from to build the quadratic model.
+        n_sample_points: The number of points to sample in each iteration.
+        n_sample_points_coeff: This specifies the number of points to sample
```
[the remark this answers is not in the document]
Comment (kevinsung) [reply recirq/optimize/mgd.py]: Done.

File: recirq/optimize/mgd.py [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+            in each iteration as a coefficient of the number of points
+            required to exactly determine a quadratic model. The number
+            of sample points will be this coefficient times (n+1)(n+2)/2,
+            rounded up, where n is the number of parameters.
+            Setting this overrides n_sample_points.
+        rate_decay_exponent: Controls decay of learning rate.
+        stability_constant: Affects decay of learning rate.
```
[the remark this answers is not in the document]
Comment (kevinsung) [reply recirq/optimize/mgd.py:122]: Done.

File: recirq/optimize/mgd.py [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+    converged = False
+    message = None
+
+    while num_iter < max_iterations:
+        current_sample_radius = (sample_radius /
+                                 (num_iter + 1)**sample_radius_decay_exponent)
+
+        # Determine points to evaluate
+        # in ball around current point
+        new_xs = [np.copy(current_x)] + [
```
[the remark this answers is not in the document]
Comment (kevinsung) [reply recirq/optimize/mgd.py:180]: `current_x` is mutated near the end of the loop. If I didn't copy it, then `known_xs` would get inadvertently mutated as well.

File: recirq/optimize/mgd_test.py [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numpy as np
+
+from recirq.optimize.mgd import model_gradient_descent
+
+
+def sum_of_squares(x):
+    return np.sum(x**2)
```
[the remark this answers is not in the document]
Comment (kevinsung) [reply recirq/optimize/mgd_test.py]: Done.

File: recirq/optimize/mgd_test.py (hunk 2 above)
Comment (kevinsung) [reply recirq/optimize/mgd_test.py:54]: Done.

File: recirq/optimize/minimize.py (hunk 3 above)
Comment (kevinsung) [reply recirq/optimize/minimize.py]: Done.

File: docs/qaoa/Optimization-Analysis.ipynb [hunk 12]
```diff
@@ -0,0 +1,568 @@
+{
```
Comment (mpharrigan) [inline docs/qaoa/Optimization-Analysis.ipynb:1]: nice notebook!

File: recirq/qaoa/experiments/optimization_tasks.py [hunk 13]
```diff
@@ -0,0 +1,267 @@
...
+        lowest_energy_index = np.argmin(energies)
+        lowest_energy = energies[lowest_energy_index]
+        if lowest_energy < lowest_energy_found:
+            lowest_energy_found = lowest_energy
+            best_bitstring_found = bitstrings[lowest_energy_index]
+        mean = np.mean(energies)
+        # Save bitstrings
+        nonlocal iteration
+        iteration_params = OptimizationIterationParameters(task, iteration)
+        recirq.save(task=iteration_params,
```
[the remark this answers is not in the document]
Comment (kevinsung) [reply recirq/qaoa/experiments/optimization_tasks.py:176]: I don't see a clear use-case for intermediate results. I'll just save one file at the end of the optimization.

File: recirq/qaoa/experiments/optimization_tasks.py [hunk 14]
```diff
@@ -0,0 +1,267 @@
+from typing import Dict, Sequence, Optional, List, Tuple
+from dataclasses import dataclass
```
[the remark this answers is not in the document]
Comment (kevinsung) [reply recirq/qaoa/experiments/optimization_tasks.py]: Done.

File: recirq/qaoa/experiments/optimization_tasks.py [hunk 15]
```diff
@@ -0,0 +1,267 @@
...
+    generation_task: ProblemGenerationTaskT
+    device_name: str
+    p: int
+    algorithm: OptimizationAlgorithm
+    x0: List[float]
+    line_placement_strategy: Optional[str] = None
+
+    @property
+    def dirname(self):
+        fn = os.path.join(self.dataset_id, self.device_name,
```
[the remark this answers is not in the document]
Comment (kevinsung) [reply recirq/qaoa/experiments/optimization_tasks.py]: Done.

File: recirq/qaoa/experiments/optimization_tasks.py [hunk 16]
```diff
@@ -0,0 +1,267 @@
...
+
+    @property
+    def fn(self):
+        return os.path.join(self.dirname, 'result')
+
+
+@recirq.json_serializable_dataclass(namespace='recirq.qaoa',
+                                    registry=recirq.Registry,
+                                    frozen=True)
+class OptimizationIterationParameters:
```
[the remark this answers is not in the document]
Comment (kevinsung) [reply recirq/qaoa/experiments/optimization_tasks.py]: It's deleted now that I save all data in one file.

File: recirq/qaoa/experiments/optimization_tasks.py [hunk 17]
```diff
@@ -0,0 +1,267 @@
+from typing import Dict, Sequence, Optional, List, Tuple
```
[the remark this answers is not in the document]
Comment (kevinsung) [reply recirq/qaoa/experiments/optimization_tasks.py:1]: Done.

File: docs/qaoa/Optimization-Analysis.ipynb (hunk 12 above)
Comment (kevinsung) [reply docs/qaoa/Optimization-Analysis.ipynb:1]: Thanks :)

**Comment by kevinsung** [comment]
@mpharrigan I've addressed your comments. PTAL.

**Comment by mpharrigan** [comment]
Thanks for addressing the comments! This looks good to me, but we'll merge after #52

**Comment by kevinsung** [comment]
Updating the QAOA tasks tutorial left a large diff for some reason. All I did was append some cells to the end, rerun the notebook, and then clear all outputs.

**Comment by mpharrigan** [comment]
yeah that doesn't look right; i'll have to investigate

**Comment by kevinsung** [comment]
Ok the difference between the old and new versions of the notebooks is in the indentation. The old version is more indented. I don't know what the cause for the discrepancy is. It may be the version of Jupyter used. I'm using JupyterLab version 2.2.0, but I tried Jupyter Notebook version 6.0.3 and got the same result.

**Comment by mpharrigan** [comment]
`nbfmt` from https://github.com/tensorflow/docs/tree/master/tools/tensorflow_docs/tools uses two spaces by default in jupyter notebooks. I think this is supposed to match google colab's style? Kindof annoying they don't just be normal. For this PR, I'll run it through to turn to 2 spaces to reduce the diff and then later we can decide which convention we want.

**Comment by mpharrigan** [comment]
Merged in #65 , so diff should calm way down after that gets in. I've kept two-json-indents for `tasks.ipynb` but one for `Optimization-Analysis.ipynb`. The colab frontmatter and renaming to be lowercase will happen in a follow-on PR

**Pull request merged**